main.go

```diff
@@ -40,22 +40,30 @@ func main() {
 		fmt.Printf("%s\n", config.GetDefaultConfig())
 		os.Exit(0)
 	}
+
 	appConfig, err := config.NewAppConfig("lazygit", version, commit, date, buildSource, *debuggingFlag)
 	if err != nil {
 		log.Fatal(err.Error())
 	}
 
-	app, err := app.NewApp(appConfig)
-
-	if err == nil {
-		err = app.Run()
+	appDir := flag.Arg(0)
```

Comment:
does this still work if we also pass in a named flag like '-v'?

Reply:
Just tested the code:
- `lg -v` shows version
- `lg someGitDir` opens lg in someGitDir folder
- `lg -v someGitDir` shows version
- `lg someGitDir -v` opens lg in someGitDir folder

I don't think this is the right behavior.  
Personally i would prefer that when the `-v` flag is specified it always shows the version.

Reply:
Agreed, that seems to be an unwanted side effect, will update it now.

Reply:
It appears that when you pass `dir -v` the flags package thinks that `-v` is an argument. This seems to be an issue with the flags package, we could change to some library like github.com/spf12/cobra, or change the way the flags are interpreted on our side. I'd prefer to use something like cobra.

Reply:
It's one up spf13 not spf12 :),  
The library seems fine to me although from first looks it looks like quite a big library for our use.  
For some alternatives here is a list with CLI libraries: https://github.com/avelino/awesome-go#standard-cli

Reply:
Typo :D // Fixed it, it's indeed a bit on the large side, but there are many features that we could use. Including the integration of viper.

Reply:
If lg can make use of those features the usage of that package seems fine to me.

Reply:
The features I would think we could use is commands like what git itself has, and viper would be ideal for flag parsing and the configuration of keybinds / color schemes, like what https://github.com/wagoodman/dive/blob/master/cmd/root.go has.

Reply:
@jesseduffield Do you have any idea on how i could solve this problem?

Reply:
I had a look at that list of packages and flaggy looks like the right choice. It's got support for positional arguments, and doesn't require any refactoring. Here's a patch I just whipped up which satisfies our use case.

If think we want to hold off on Cobra until we have a strong use case for using it: it looks pretty heavy and flaggy seems to be fine for now (it also has a nice --help option out of the box)

```
diff --git a/main.go b/main.go
index 09ed8b2..bcee464 100644
--- a/main.go
+++ b/main.go
@@ -1,7 +1,6 @@
 package main
 
 import (
-	"flag"
 	"fmt"
 	"log"
 	"os"
@@ -9,6 +8,7 @@ import (
 	"runtime"
 
 	"github.com/go-errors/errors"
+	"github.com/integrii/flaggy"
 	"github.com/jesseduffield/lazygit/pkg/app"
 	"github.com/jesseduffield/lazygit/pkg/config"
 )
@@ -18,10 +18,6 @@ var (
 	version     = "unversioned"
 	date        string
 	buildSource = "unknown"
-
-	configFlag    = flag.Bool("config", false, "Print the current default config")
-	debuggingFlag = flag.Bool("debug", false, "a boolean")
-	versionFlag   = flag.Bool("v", false, "Print the current version")
 )
 
 func projectPath(path string) string {
@@ -30,27 +26,41 @@ func projectPath(path string) string {
 }
 
 func main() {
-	flag.Parse()
-	if *versionFlag {
+	flaggy.DefaultParser.ShowVersionWithVersionFlag = false
+
+	repoPath := "."
+	flaggy.AddPositionalValue(&repoPath, "path", 1, false, "Path of git repo")
+
+	versionFlag := false
+	flaggy.Bool(&versionFlag, "v", "version", "Print the current version")
+
+	debuggingFlag := false
+	flaggy.Bool(&debuggingFlag, "d", "debug", "Run in debug mode with logging")
+
+	configFlag := false
+	flaggy.Bool(&configFlag, "c", "config", "Print the current default config")
+
+	flaggy.Parse()
+
+	if versionFlag {
 		fmt.Printf("commit=%s, build date=%s, build source=%s, version=%s, os=%s, arch=%s\n", commit, date, buildSource, version, runtime.GOOS, runtime.GOARCH)
 		os.Exit(0)
 	}
 
-	if *configFlag {
+	if configFlag {
 		fmt.Printf("%s\n", config.GetDefaultConfig())
 		os.Exit(0)
 	}
 
-	appConfig, err := config.NewAppConfig("lazygit", version, commit, date, buildSource, *debuggingFlag)
+	appConfig, err := config.NewAppConfig("lazygit", version, commit, date, buildSource, debuggingFlag)
 	if err != nil {
 		log.Fatal(err.Error())
 	}
 
-	appDir := flag.Arg(0)
-	if appDir != "" {
-		err := os.Chdir(appDir)
+	if repoPath != "." {
+		err := os.Chdir(repoPath)
 		if err != nil {
-			log.Fatalf("Could not change directory to %s", appDir)
+			log.Fatalf("Could not change directory to %s", repoPath)
 		}
 	}
 

```

Reply:
I think the help message from flaggy is rather messy compared to others in the list. Mostly the sorting and spacing is on the dirty side. Cobra has the advantage that it is used in a lot of programs so the testing is done and the structure the help message across those programs is all the same. Also cobra being from the same author as viper means the integration with viper is rather easy which means our road to configurability is likely to benefit from closely connected tool "stack". I havent done any testing with benchmarks to comment on the speed / size. But i dont think thats something we will trip over.

Reply:
I just checked, lazygit is about 25MB on my machine.
If i create a simple hello world program with log.Printf it is compiled to a 2.1MB binary.
If i create a simple hello world program with log.Printf and cobra, it is compiled to a 4.1MB binary.
Sounds like a lot but i still think it is not much compared the 25MB.
About the performance, since cobra is run a startup we would likely not notice it

Reply:
Apologies for the delay in responding, I've been having an impromptu break :)

I think we should go for flaggy for now. I don't think the use case for cobra is really there, and I prefer to use plug-and-play libraries over frameworks that require refactoring.